src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs

```diff
@@ -260,50 +260,58 @@ private Dispatcher Dispatcher
 
         private void SubmissionBufferAdded(object sender, SubmissionBufferAddedEventArgs args)
         {
-            AddSubmission(_currentWindow.TextView, args.NewBuffer);
+            AddSubmission(_currentWindow.TextView, args.NewBuffer, this.LanguageName);
         }
 
         // The REPL window might change content type to host command content type (when a host command is typed at the beginning of the buffer).
         private void LanguageBufferContentTypeChanged(object sender, ContentTypeChangedEventArgs e)
         {
             var buffer = e.Before.TextBuffer;
-            var afterIsCSharp = e.AfterContentType.IsOfType(ContentTypeNames.CSharpContentType);
-            var beforeIsCSharp = e.BeforeContentType.IsOfType(ContentTypeNames.CSharpContentType);
 
-            if (afterIsCSharp == beforeIsCSharp)
+            var afterIsLanguage = e.AfterContentType.IsOfType(this.ContentType.TypeName);
+            var afterIsInteractiveCommand = e.AfterContentType.IsOfType(PredefinedInteractiveCommandsContentTypes.InteractiveCommandContentTypeName);
+            var beforeIsLanguage = e.BeforeContentType.IsOfType(this.ContentType.TypeName);
+            var beforeIsInteractiveCommand = e.BeforeContentType.IsOfType(PredefinedInteractiveCommandsContentTypes.InteractiveCommandContentTypeName);
+
+            if (afterIsLanguage == beforeIsLanguage ||
```

Comment:
@ManishJayaswal, that condition is already below (beforeIsInteractiveCommand != afterIsInteractiveCommand). The condition above was already in the code and is also legitimate -- i.e. we shouldn't carry on ripping out the submission project and creating a new one if the buffer is just changing from C# to C#.

FWIW, I think @jasonmalinowski is correct. If we want to be paranoid here, we should probably only move forward if e.AfterContentType != e.BeforeContentType. It's extremely unlikely this will ever happen. (I suspect the editor won't fire the event unless the content type actually changes :smile:). However, we can leave the assert below to verify that the content types have the values we expect.